### PR TITLE
feat: add gatekeeper plugin for auto-approve safe commands

### PR DESCRIPTION
**Pull request by amondnet**
## Summary

Adds the **gatekeeper plugin**, a two-layer security system for Claude Code that automatically approves safe commands and blocks dangerous operations while delegating unknown commands to an AI security agent for review.

### Layer 1: Pattern Matching (PreToolUse Hook)
- **Instant allow**: Safe read-only commands (ls, cat, git status, etc.)
- **Instant deny**: Dangerous operations (rm -rf, git push --force, etc.)
- **Unknown commands**: Pass to Layer 2

### Layer 2: AI Security Review (PermissionRequest Hook)
- AI agent analyzes command intent and risk level
- Provides security recommendations
- Learns from user patterns for better future decisions

## Key Features

- ✅ Zero-friction for safe operations
- ✅ Protection against accidental destructive commands
- ✅ Intelligent unknown command review
- ✅ Extensible pattern configuration
- ✅ Full test coverage

## Changes

- Added `plugins/gatekeeper/` directory with complete plugin implementation
- Added gatekeeper entry to marketplace.json
- Updated workspace configuration to include plugins/*

## Plugin Structure

```
plugins/gatekeeper/
├── .claude-plugin/
│   └── plugin.json         # Plugin manifest
├── hooks/
│   └── hooks.json          # PreToolUse hook configuration
├── src/
│   ├── pre-tool-use.ts     # Pattern matching logic
│   └── pre-tool-use.test.ts # Test suite
├── README.md               # Plugin documentation
└── package.json            # Dependencies
```

## Test Plan

- [x] Unit tests for pattern matching (allow/deny/unknown)
- [x] Test coverage for all command patterns
- [x] Integration test with PreToolUse hook
- [x] Verify hook registration in plugin.json

## Related Documentation

- See `plugins/gatekeeper/README.md` for usage guide
- See docs/plugins.md for Claude Code plugin architecture
- See docs/lessons-learned/ for plugin development best practices

### CODE REVIEW

File: plugins/gatekeeper/src/pre-tool-use.ts [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+  },
+]
+
+export const ALLOW_RULES: Rule[] = [
+  // Package managers
+  {
+    pattern:
+      /^(npm|yarn|pnpm|bun)\s+(test|run|install|ci|add|remove|exec|ls|info|outdated|audit|why|x)\b/i,
+    reason: 'Safe package manager command',
+  },
```
[the remark this answers is not in the document]
Comment (amondnet) [reply plugins/gatekeeper/src/pre-tool-use.ts]: ✅ Applied in 5a7b852
- File: `plugins/gatekeeper/src/pre-tool-use.ts`
- Change: Removed `exec` and `x` from package manager allowlist. These subcommands now passthrough to AI review.

Thanks for the suggestion, @gemini-code-assist!

File: plugins/gatekeeper/src/pre-tool-use.ts [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+  // Docker read operations
+  {
+    pattern: /^docker\s+(ps|logs|images|inspect|version)\b/i,
+    reason: 'Safe docker read operation',
+  },
+]
+
+export function isGitPushNonForce(cmd: string): boolean {
+  return /^git\s+push\b/i.test(cmd) && !/--force\b|-f\b/i.test(cmd)
+}
```
[the remark this answers is not in the document]
Comment (amondnet) [reply plugins/gatekeeper/src/pre-tool-use.ts]: ✅ Applied in 5a7b852
- File: `plugins/gatekeeper/src/pre-tool-use.ts`
- Change: Updated `isGitPushNonForce` regex to detect combined short flags like `-vf` using `/--force(?:-with-lease)?\b|\s-[^\s]*f/i`

Thanks for the suggestion, @gemini-code-assist!

File: plugins/gatekeeper/src/pre-tool-use.ts [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+export const DENY_RULES: Rule[] = [
+  { pattern: /^rm\s+-rf\s+\/(?:\s|$)/i, reason: 'Filesystem root deletion blocked' },
+  { pattern: /^rm\s+-rf\s+~(?:\/|$)/i, reason: 'Home directory deletion blocked' },
+  { pattern: /^mkfs\./i, reason: 'Disk format command blocked' },
+  {
+    pattern: /^dd\s+if=\/dev\/zero\s+of=\/dev\//i,
+    reason: 'Disk zeroing blocked',
+  },
+]
```
[the remark this answers is not in the document]
Comment (amondnet) [reply plugins/gatekeeper/src/pre-tool-use.ts]: ✅ Applied in 5a7b852
- File: `plugins/gatekeeper/src/pre-tool-use.ts`
- Change: Added `rm -rf /*` pattern to DENY_RULES with reason 'Destructive wildcard deletion from root blocked'

Thanks for the suggestion, @gemini-code-assist!

File: plugins/gatekeeper/src/pre-tool-use.ts [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+} from '@anthropic-ai/claude-agent-sdk'
+import process from 'node:process'
+
+export interface Rule {
+  pattern: RegExp
+  reason: string
+}
+
+export const DENY_RULES: Rule[] = [
+  { pattern: /^rm\s+-rf\s+\/(?:\s|$)/i, reason: 'Filesystem root deletion blocked' },
```
[the remark this answers is not in the document]
Comment (amondnet) [reply plugins/gatekeeper/src/pre-tool-use.ts]: ⏭️ Not applied
- Reason: Current passthrough behavior for commands with leading whitespace is safer — they go to AI review instead of being auto-approved. Adding `\s*` would auto-approve these edge cases.
- Alternative: Commands with leading whitespace continue to passthrough to AI review as before.

Thank you for understanding, @gemini-code-assist!

File: plugins/gatekeeper/src/pre-tool-use.ts [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+  // Build tools and runtimes
+  {
+    pattern:
+      /^(node|npx|tsx|python3?|ruby|go\s+run|cargo\s+(build|run|test|check|clippy)|make|gradle|mvn)\b/i,
+    reason: 'Safe build/runtime command',
+  },
+  // File inspection commands
+  {
+    pattern:
+      /^(ls|pwd|cat|head|tail|wc|file|which|type|env|echo|printf|grep|find|rg|fd|ag|tree)\b/i,
```
[the remark this answers is not in the document]
Comment (amondnet) [reply plugins/gatekeeper/src/pre-tool-use.ts]: ✅ Applied in 5a7b852
- File: `plugins/gatekeeper/src/pre-tool-use.ts`
- Change: Added command chaining/substitution detection before DENY/ALLOW checks. Commands containing `;`, `&&`, `||`, `|`, backticks, or `$()` now passthrough to AI review instead of being auto-approved.
- Tests updated to reflect the new behavior.

Thanks for the thorough analysis, @cubic-dev-ai!

File: plugins/gatekeeper/src/pre-tool-use.test.ts [hunk 6]
```diff
@@ -0,0 +1,393 @@
...
+    expectDeny(bash('rm -rf /'))
+    // "mkfs.ext4" is denied even if prefixed by sudo (which could look like a command)
+    expectDeny(bash('mkfs.ext4 /dev/sda'))
+  })
+})
+
+// ─── Edge cases ──────────────────────────────────────────────────────────────
+
+describe('edge cases', () => {
+  test('should allow chained commands if first command matches allow pattern', () => {
```
[the remark this answers is not in the document]
Comment (amondnet) [reply plugins/gatekeeper/src/pre-tool-use.test.ts]: 💡 Implemented with alternative approach (5a7b852)
- Suggestion: Add a test documenting that chained safe+dangerous commands are auto-allowed
- Applied: Instead of documenting the vulnerability, we fixed it. Chained commands now passthrough to AI review. Tests updated to verify this new secure behavior.
- Reason: Better to fix the root cause than document the limitation

Thanks for flagging this, @cubic-dev-ai!

File: plugins/gatekeeper/.gitignore [hunk 7]
```diff
@@ -0,0 +1 @@
+!dist/
```
[the remark this answers is not in the document]
Comment (amondnet) [reply plugins/gatekeeper/.gitignore]: ✅ Applied in 5a7b852
- File: `plugins/gatekeeper/.gitignore`
- Change: Added `!dist/**` to also unignore directory contents, not just the directory itself.

Thanks for the catch, @cubic-dev-ai!

File: plugins/gatekeeper/src/pre-tool-use.ts [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+  },
+  // Docker read operations
+  {
+    pattern: /^docker\s+(ps|logs|images|inspect|version)\b/i,
+    reason: 'Safe docker read operation',
+  },
+]
+
+export function isGitPushNonForce(cmd: string): boolean {
+  return /^git\s+push\b/i.test(cmd) && !/--force(?:-with-lease)?\b|\s-[^\s]*f/i.test(cmd)
```
[the remark this answers is not in the document]
Comment (amondnet) [reply plugins/gatekeeper/src/pre-tool-use.ts]: ✅ Applied in b30980a
- File: `plugins/gatekeeper/src/pre-tool-use.ts`
- Change: Updated regex to `\s-(?!-)\S*f` using negative lookahead to only match single-dash flags. `--follow-tags` and `--no-verify` are now correctly allowed.
- Tests added for both false positive cases.

Thanks for the thorough regex analysis, @cubic-dev-ai!
